### PR TITLE
[WIP] Implement DefaultQubitTorch prototype

### PR DESCRIPTION
**Pull request by iamlucaswolf**
**Context:**

This is a (early) prototype of a `DefaultQubitTorch`  "passthru" device for end-to-end GPU-bound backpropagation with PyTorch (similar to `pennylane.devices.default_qubit_tf.DefaultQubitTF). The idea came from earlier discussion with @josh146 on issue #919

*I'm aware that this PR is still incomplete and cannot (and should not) be merged as is, I just wanted your feedback on a few points.* 

Currently, analytic measurements of observables (expval, var) seem to work (although thorough testing is still to be done). However, I'm still undecided with the following issues; your input would be greatly appreciated:

 a) **_Sampling_**
Sampling in `pennylane.QubitDevice` still gives me a bit of a headache. Executing a circuit with "analytic" observables returns `torch.tensor`, whereas the sampling methods of `QubitDevice` (i.e. `generate_samples()`, `sample_basis_states()`, `generate_basis_states()`, ...) all return `np.array`. This should not be that big of a deal per se, as we cannot backprop through sampling anyway -- however it results in an inconsistent interface. This may cause havoc downstream (e.g. when some other functionality relies on compatible types), since `numpy.array` and `torch.tensor` don't always play well together.
Another aspect that ties into to this is `QubitDevice.execute()` line 208ff. I don't think that I have entirely understood what's going on there yet, however, packing results in an `np.array(..., dtype='object')` becomes a problem when converting results into `torch.tensor`, since PyTorch doesn't have an "object" dtype. Do you have an idea how this could be resolve?

b) **_Complex numbers in PyTorch_**
Support for complex numbers in PyTorch is relatively new, and therefore some operations (e.g. `torch.norm`) are not yet implemented. Most annoyingly, this includes `torch.einsum` which seems to work for _some_ indices (as in `DefaultQubit._apply_diagonal_unitary()`), but not for others (specifically, in `DefaultQubit._apply_unitary_einsum()`). For this reason, I patched the latter in `DefaultQubitTorch` to simply fallback to `_apply_unitary()`. I'm not sure if anything can be done about this other than waiting for FAIR to fix these things; I just wanted to point out that this is the reason for some workarounds.

c) **_Other things to keep in mind_**
I'm still new to PennyLane, so I don't know all of the codebase. Are there other parts that might interfere with my approach?

d) **_Helper modules_**
`torch_ops.py` and `tf_ops.py` are very similar. Perhaps the two could be refactored/merged. Yay or nay?

**Description of the Change:**
See above.

This PR is just a minimum viable prototype. Refactoring, testing, and documentation are still on the to-do list.

**Benefits:**
See above.

**Possible Drawbacks:**
Added complexity.

**Related GitHub Issues:**
#919 

### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @iamlucaswolf! This is a great contribution -- PyTorch backpropagation has been high on our wish list, ever since PyTorch began supporting complex arithmetic.

> Sampling in pennylane.QubitDevice still gives me a bit of a headache. Executing a circuit with "analytic" observables returns torch.tensor, whereas the sampling methods of QubitDevice (i.e. generate_samples(), sample_basis_states(), generate_basis_states(), ...) all return np.array

This is very good point. I had a look at `default.qubit.tf` to see what happens, and it appears that the entire pipeline from the point of generating samples is using NumPy arrays. While this is converted back to a TensorFlow array at the last minute; within `QubitDevice.execute()` in the line
```python
results = self._asarray(results)
```
this is not ideal.

I'm not sure what the best solution here is.

* We could make the generic sampling functionality in `QubitDevice` framework agnostic (as we have been slowly doing with `QubitDevice`). A potential downside is increased maintenance and complexity.

* We could have `default.qubit.tf` and `default.qubit.torch` simply overwrite the sampling methods where needed. Perhaps this is a better approach, it lends to cleaner code and gives more flexibility.

* Finally, a fallback could simply be to have the finite-shot simulation use NumPy for sampling, but return the output as a Torch tensor (like the TF simulator currently does). Not the best solution, but gives us time to come up with a better one.

> I don't think that I have entirely understood what's going on there yet, however, packing results in an np.array(..., dtype='object') becomes a problem when converting results into torch.tensor, since PyTorch doesn't have an "object" dtype. Do you have an idea how this could be resolve?

Ah - the reason behind this line, is that it is sometimes possible to construct a QNode that returns ragged/jagged output if sampling and measurement statistics are combined: 

For example:

```python
dev = qml.device("default.qubit.autograd", wires=2, shots=5, analytic=False)

@qml.qnode(dev, interface="autograd", diff_method="backprop")
def circuit(weights):
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
    return qml.sample(qml.PauliZ(0)), qml.expval(qml.PauliX(1))
```

returns 

```python
>>> weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
>>> print(circuit(weights))
[tensor([1, 1, 1, 1, 1], requires_grad=True)
 tensor(-0.2, requires_grad=True)]
```

The reason we return an object array here is that we decided (at some point) that the output type of a QNode should not change depending on the measurement type - if it is usually an array (in this example), it should continue to be an array.

However, I think this is likely the wrong approach. I noticed also that this particular line (`self._asarray(..., dtype='object')`) *also* fails for the `default.qubit.tf` simulator, since TensorFlow has no `object` dtype.

Perhaps the solution here (which is a more general change required in PL behaviour):

* If the QNode output is a mixture of sampling and measurement statistics, simply return a list of arrays/tensors, rather than attempting to output a ragged array/tensor.

> Most annoyingly, this includes torch.einsum which seems to work for some indices (as in DefaultQubit._apply_diagonal_unitary()), but not for others (specifically, in DefaultQubit._apply_unitary_einsum()). For this reason, I patched the latter in DefaultQubitTorch to simply fallback to _apply_unitary().

It's a shame that there is not `torch.einsum` complex support yet! It is fine to simply fallback to `_apply_unitary` for now 🙂 

>  torch_ops.py and tf_ops.py are very similar. Perhaps the two could be refactored/merged. Yay or nay?

This is a very good question (we also have `autograd_ops.py`, for `default.qubit.autograd`!

I would suggest simply leave it for now. We are currently working on implementing a unified API for framework-agnostic tensor manipulations (see the [`qml.proc`](https://pennylane.readthedocs.io/en/latest/code/qml_proc.html) experimental module, and #922). This is perhaps a prime candidate for a re-write using `qml.proc` once it is more mature/stable.

**Comment by iamlucaswolf** [comment]
Thanks for your response, @josh146 !

> We could have default.qubit.tf and default.qubit.torch simply overwrite the sampling methods where needed. Perhaps this is a better approach, it lends to cleaner code and gives more flexibility.

This appears to me to be the cleanest option? I have to catch up on some coursework today, but hopefully will be able to explore this idea tonight/over the next few days. I'll come back to you one I have a solution working for `DefaultQubitTorch` (that hopefully extrapolates well to `DefaultQubitTF`).

> Ah - the reason behind this line, is that it is sometimes possible to construct a QNode that returns ragged/jagged output if sampling and measurement statistics are combined:
>
> [...]
>
> The reason we return an object array here is that we decided (at some point) that the output type of a QNode should not change depending on the measurement type - if it is usually an array (in this example), it should continue to be an array. 

Ah, that makes sense.

> Perhaps the solution here (which is a more general change required in PL behaviour):
>
 > -   If the QNode output is a mixture of sampling and measurement statistics, simply return a list of arrays/tensors, rather than attempting to output a ragged array/tensor.

IMHO the most general/flexible approach would be to return a dict that maps wires to measurements where values are scalars for statistics and arrays/tensors for samples. For example, consider:

```python
dev = qml.device(...)

@qml.qnode(dev)
def circuit(x,y):
    qml.CNOT(wires=[0,1])
    qml.CNOT(wires=[1,2])
    qml.RY(x, wires=0)
    qml.RX(y, wires=2)
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
```

```python
>>> circuit(0.5, 0.2)
{ '0':  0.858, '2': 0.982}
```

This method comes with a few extra benefits:

 - It's immediately clear which qubits were measured. Currently, this information gets lost, which might result in subtle bugs when indexing the output array (e.g. for the above circuit, `res = circuit(0.5, 0.2); res[1]` would yield the measurement for qubit **2**, not 1, under the current API)
 - For PyTorch, only those measurements whose computation depends on tensors with `requires_grad=True` would have their `grad_fn` set

To restore the current behaviour, we could also add a kwarg `combine` to `BaseQNode.__call__` that's passed down to the device's `execute()` method in order to stack all measurements into a single array if all measurements are statistics (and throw a`ValueError` otherwise)

```python
>>> circuit(0.5, 0.2, combine=True)
array([0.858, 0.982])
```

That being said, I don't know how far reaching these changes would be (I guess all the plugin devices would be affected by that as well?) Also, this will break backwards compatibility. 

What are your thoughts on this?

**Comment by josh146** [comment]
> This method comes with a few extra benefits:
> * It's immediately clear which qubits were measured. Currently, this information gets lost, which might result in subtle bugs when indexing the output array (e.g. for the above circuit, res = circuit(0.5, 0.2); res[1] would yield the measurement for qubit 2, not 1, under the current API)
> * For PyTorch, only those measurements whose computation depends on tensors with requires_grad=True would have their grad_fn set

This is an interesting idea, definitely. I find the second bullet point much more convincing --- I think there is a huge advantage to being able to return non-differentiable samples _and_ differentiable measurement statistics in the same call.

> That being said, I don't know how far reaching these changes would be (I guess all the plugin devices would be affected by that as well?) Also, this will break backwards compatibility.

Yes, unfortunately a change of this scale requires a bit more exploration and design work! Too large to include it alongside a new Torch simulator. I would suggest:

1. We implement a minimal working Torch simulator subject to current PennyLane constraints

2. We then re-evaluate existing PennyLane constraints

3. Updates and improvements to the simulators resulting from (2) can be made.

> To restore the current behaviour, we could also add a kwarg combine to BaseQNode.__call__ 

I just want to point out that we are in the process of refactoring PennyLane's internals to unblock a lot of exciting future features. Due to the scope of the change, we are currently supporting two 'cores' -- the original core (`BaseQNode`), and the new core, referred to as [tape mode](https://pennylane.readthedocs.io/en/latest/code/qml_tape.html).

At the moment, tape mode more-or-less has the same external behaviour as non-tape mode (with many new features). However, we do have the freedom to introduce non-backwards compatible changes into tape mode if we determine that they lead to improvements.

**Comment by iamlucaswolf** [comment]
A few updates:

a) **PyTorch complex**

I realized that `torch.tensordot` and `torch.stack` don't support backprop when used with complex tensors. This is really annoying, as it renders the `DefaultQubitTorch` class more or less useless in its current state. 

I am not sure whether/how these functions could be implemented manually, and at this point I think it would be wiser to just wait until torch finally supports `complex128` as a first-class citizen. Maybe @josh146 has some thoughts on this?

b) **Return types and sampling**

I experimented a bit with making the sampling procedure in `QubitDevice` framework-independent. I think this should be doable in principal, however it would require a number of changes that probably warrant their own PR.

Instead, I now have a version where everything related to sampling is done in NumPy and converted into the respective array representation just before `execute` returns (basically, how the existing NumPy/Tensorflow devices already work). Mixed measurements (e.g. `expval` and `sample` in one circuit) are still a problem in  `DefaultQubitTF` and `DefaultQubitTorch`, however, as discussed above, a clean fix would require changing the existing API. Therefore, the current return types are as follows (for both analytic and sample-based measurements):

| Measurement        | `DefaultQubit`           | `DefaultQubitTF`  | `DefaultQubitTorch`
| ------------- |:-------------:|:-----:|:---:|
| `expval`, `var`, `prob` | `np.ndarray`, `dtype=np.float64` |  `tf.Tensor`, `dtype=tf.float64` | `torch.Tensor`, `dtype=torch.float64` |
| `sample` | `np.ndarray`, `dtype=np.int64` | `tf.Tensor`, `dtype=tf.int64` | `torch.Tensor`, `dtype=torch.int64` | 
| mixed (e.g. `expval` and `sample`) | `np.ndarray`, `dtype='object'` | fails | fails |

c) **Going forward**

I'd  say that we leave this PR open until torch has sufficient `complex128` support. In the meantime, I'd add documentation and do some refactoring w.r.t coding style. Thoughts?

**Comment by josh146** [comment]
> I realized that torch.tensordot and torch.stack don't support backprop when used with complex tensors. This is really annoying, as it renders the DefaultQubitTorch class more or less useless in its current state.

This is very annoying. 

> I am not sure whether/how these functions could be implemented manually, and at this point I think it would be wiser to just wait until torch finally supports complex128 as a first-class citizen. Maybe @josh146 has some thoughts on this?

I think this is a good idea, hopefully better complex feature parity is available in Torch version 1.8.

Since Torch now at least has a complex type, we _could_ have the simulator accept and return complex numbers, but perform the bulk of the simulation using purely real numbers; that is, [representing complex numbers as real 2x2 matrices](https://en.wikipedia.org/wiki/Complex_number#Matrix_representation_of_complex_numbers). However, this will come at the expense of additional overhead, so probably is not worth it.

> Instead, I now have a version where everything related to sampling is done in NumPy and converted into the respective array representation just before execute returns (basically, how the existing NumPy/Tensorflow devices already work).

This should be sufficient for the near-term :+1:

In the medium term, we have been playing with the idea of tailoring the device choice to the QNode that is being created. For example,

* `default.qubit` with `interface="tf", analytic=True`: automatically return `default.qubit.tf` for backpropagation
* `default.qubit` with `interface="tf", analytic=False`: automatically return `default.qubit` for the parameter-shift rule.
* etc.

In this case, there might not be so much of a need to preserve TF datatypes during sampling? Perhaps there are exceptions though, for example, a large simulation using TF on GPUs, where the gradient is not required.

> I'd say that we leave this PR open until torch has sufficient complex128 support. In the meantime, I'd add documentation and do some refactoring w.r.t coding style. Thoughts?

That would be great @iamlucaswolf! Even though we are still waiting on full complex support, this PR already provides a solid base  for the finished simulator, as well as uncovering existing bugs/issues. Solid work 💪

**Comment by shukob** [comment]
Hi, I tested this and the `jacobian` operation does not work.

First, I corrected the typo `passthru_intergace="torch",` to `passthru_interface="torch",` in `default_qubit_torch.py` L131, then added `       'default.qubit.torch = pennylane.devices.default_qubit_torch:DefaultQubitTorch'` in `setup.py`. With the installation of this version of the pennylane, the following error occurred.

```
import pennylane as qml
import torch
from pennylane import QNodeCollection
n_qubits = 4
qml.enable_tape()
dev = qml.device("default.qubit.torch", wires=n_qubits)
@qml.qnode(dev, interface='torch', diff_method='backprop')
def circuit(inputs, weights):
    qml.templates.AngleEmbedding(inputs, wires=range(n_qubits))
    qml.templates.BasicEntanglerLayers(weights, wires=range(n_qubits))
    return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]

n_layers = 1
weight_shapes = {"weights": (n_layers, n_qubits)}

qlayer = qml.qnn.TorchLayer(circuit, weight_shapes)

x = torch.rand((4, 1))
norm = x.norm(p=2, dim=1, keepdim=True)
x_normalized = x.div(norm).to(torch.device('cuda'))
val = qlayer(x_normalized)
loss = torch.sum(val).squeeze()

loss.backward()

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

<ipython-input-3-fc86e564a89c> in <module>
     16 loss = torch.sum(val).squeeze()
     17 
---> 18 loss.backward()
     19 
     20 

~/anaconda3/envs/***/lib/python3.7/site-packages/torch/tensor.py in backward(self, gradient, retain_graph, create_graph)
    219                 retain_graph=retain_graph,
    220                 create_graph=create_graph)
--> 221         torch.autograd.backward(self, gradient, retain_graph, create_graph)
    222 
    223     def register_hook(self, hook):

~/anaconda3/envs/***/lib/python3.7/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
    130     Variable._execution_engine.run_backward(
    131         tensors, grad_tensors_, retain_graph, create_graph,
--> 132         allow_unreachable=True)  # allow_unreachable flag
    133 
    134 

~/anaconda3/envs/***/lib/python3.7/site-packages/torch/autograd/function.py in apply(self, *args)
     87     def apply(self, *args):
     88         # _forward_cls is defined by derived class
---> 89         return self._forward_cls.backward(self, *args)  # type: ignore
     90 
     91 

~/anaconda3/envs/***/lib/python3.7/site-packages/PennyLane-0.14.0.dev0-py3.7.egg/pennylane/tape/interfaces/torch.py in backward(ctx, grad_output)
     73 
     74         tape.set_parameters(ctx.all_params_unwrapped, trainable_only=False)
---> 75         jacobian = tape.jacobian(device, params=ctx.args, **tape.jacobian_options)
     76         tape.set_parameters(ctx.all_params, trainable_only=False)
     77 

~/anaconda3/envs/***/lib/python3.7/site-packages/PennyLane-0.14.0.dev0-py3.7.egg/pennylane/tape/tapes/jacobian_tape.py in jacobian(self, device, params, **options)
    454 
    455         if method not in ("best", "numeric", "analytic", "device"):
--> 456             raise ValueError(f"Unknown gradient method '{method}'")
    457 
    458         if params is None:

ValueError: Unknown gradient method 'backprop'
```

I am investigating what is happening.

**Comment by josh146** [comment]
> Also, this might be off-topic, but both TF and Torch interfaces convert tensors to NumPy array back and force. I wonder if this will be a bottleneck if the whole differentiable pipeline is build using TF/Torch.

Hey @shukob! Just to answer this question specifically quickly:

The interfaces are *only* for custom gradient registration, that is, when we want to register our own, custom, gradient rule like parameter-shift or finite difference. In these cases, the QNode is like a blackbox -- all input tensors are unwrapped to NumPy arrays.

In `"backprop"` mode, a custom gradient interface is not needed at all, PyTorch or TensorFlow owns the entire computation, and no unwrapping occurs; the tensors remain throughout.

The interface *argument* in the QNode however remains, only as a UI element. This was partially based on user feedback; to most PennyLane users, the 'interface' argument simply denotes that the QNode integrates with that framework.

For example:

```python
# The TF interface, with tensor unwrapping and custom gradient registration
# is added to the created QNode. QNode is a blackbox.
qml.QNode(func, dev_tf, interface="tf", diff_method="parameter-shift")

# The interface argument doesn't do/apply/change anything. It is there simply
# for UX consistency. There is also useful validation, so that interface='torch' and
# diff_method='backprop' with a TF device will raise an error.
qml.QNode(func, dev_tf, interface="tf", diff_method="backprop")
```

**Comment by shukob** [comment]
@josh146 Thank you for the explanation! So the execution path when using `torch` device with `interface="torch"` and `diff_method="backprop"` going through `forward` / `backward` in `_TorchInterface` is something that is not expected? I will check why this happens in my script.

File: pennylane/devices/default_qubit_torch.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    @staticmethod
+    def _allclose(a, b, atol=1e-08):
+        return torch.allclose(a, torch.as_tensor(b, dtype=a.dtype), atol=atol)
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            passthru_intergace="torch",
```
Comment (josh146) [inline pennylane/devices/default_qubit_torch.py:131]:
```suggestion
            passthru_interface="torch",
```

**Comment by josh146** [comment]
> So the execution path when using torch device with interface="torch" and diff_method="backprop" going through forward / backward in _TorchInterface is something that is not expected?

Yes that's right, I wouldn't expect this! 

I just had a look, and I think the reason PL is attempting to apply the Torch interface is because of this typo here:

https://github.com/PennyLaneAI/pennylane/pull/929/files#r539857026

That is, PennyLane was unable to identify this device as one that supports backprop. Now that the typo is fixed, could you give it another go and see if that fixes the issue?

**Comment by shukob** [comment]
@josh146 Thank you, but I already fixed the typo as described in , 

> First, I corrected the typo `passthru_intergace="torch",` to `passthru_interface="torch",` in `default_qubit_torch.py` L131, then added ` 'default.qubit.torch = pennylane.devices.default_qubit_torch:DefaultQubitTorch'` in `setup.py`.

Without this, `diff_method="backprop"` cannot be specified.

**Comment by josh146** [comment]
Oops, I missed that, sorry @shukob!🤦 Let me have a look and get back to you

**Comment by josh146** [comment]
@shukob I investigated further, and your original comment is pretty much on the mark. In particular:

* This current WIP torch device currently does not support backpropagation, as `torch.tensordot` and `torch.einsum` do not support backpropagation with complex numbers (see https://github.com/PennyLaneAI/pennylane/pull/929#issuecomment-735442000 for more discussion).

* Another complication is that `qnn.TorchLayer` currently *also* does not support backpropagation; this is why it is attempting to apply the torch interface when it shouldn't. Once this device is ready, we will need to make a separate PR to enable backprop support in `TorchLayer`, like we did for the `KerasLayer` here: https://github.com/PennyLaneAI/pennylane/pull/869

**Comment by shukob** [comment]
@josh146 Thank you for the investigation! It seems that full PyTorch support is something far, so I will switch to Keras/TF for a while.

**Comment by iamlucaswolf** [comment]
Hey @josh146 , @shukob , sorry for the silence on my part; coursework kept me busy. Also, thanks for spotting that typo!

I'll have a look at #944 sometime after the 25th meanwhile try to get info from FAIR regarding their timeline for finalizing complex support in PyTorch.

**Comment by josh146** [comment]
Update: PyTorch v1.8 has just been released, with more complex number support:

> * Complex Number support on CPU and CUDA for torch.symeig (#45121), torch.pinverse (#45819), torch.det (#45980), torch.diagflat (#47564), torch.{addcmul, addcdiv} (#46639), torch.lu_solve (#48028), torch.matrix_exp (#48363), torch.eig (#49168), torch.{acosh, asinh, atanh} (#50387), torch.masked_scatter (#51281), torch.bmm and torch.baddbmm (#42553), torch.orgqr (#50502), torch.index_fill_ (#50578), torch.cholesky_inverse (#50269)
> * Complex Number support on CUDA for torch.qr (#45032), torch.lu (#45898), torch.prod(#45980), torch.triangular_solve (#46916), torch.solve (#47045), torch.cholesky_solve (#47047), torch.mean (#47048), torch.svd (#45795), torch.inverse (#47595), torch.Tensor.index_put_ (#51148)
> * Complex Number support on CPU for torch.trace (#50380)
> * Complex Number support for torch.nn.DataParallel (#48686), torch.nn.L1Loss (#49912), Padding functions (#50594)
> * Complex Number support for torch.distributed.{all_reduce, all_gather} (#45879, #46270)
> * Complex Autograd support for torch.{atan, log, log10, log1p, log2, reciprocal, tan, pow, rsqrt, tanh, asinh, acosh} (#46275), torch.{cholesky, triangular_solve, mm, mv, ger} (#45737), torch.take(), torch.Tensor.fill_() (#46860), torch.matrix_exp (#48363), torch.{baddbmm, addbmm, addmm, addmv} (#50632), torch.qr (#48489), torch.svd and torch.pinverse (#47761), torch.sqrt (#49461), torch.diag (#51268), torch.trace (#51537), torch.exp (#47194), torch.mean (#47566), torch.addr (#50667), torch.{stack, gather, index_select}, torch.Tensor.index_add_(#49552), torch.{masked_scatter, masked_select} (#51281), torch.{addcmul, addcdiv} (#46639), torch.{acosh, asinh, atanh} (#50387), torch.solve (#47045), torch.cholesky_solve (#47047), torch.inverse (#47595)
> * Add complex autograd support for named tensors (#47289)
> * Allow converting parameters and buffers of torch.nn.Module to complex dtypes (#44788)
> * Add complex support to IValues (#50883, #51476)
> * Add TorchScript type annotation logic for complex numbers (#50884)
> * Add serialization logic for complex numbers (#51287)
> * Add support for complex number lists in JIT (#51145)
> * Add support for complex valued keys for dict in TorchScript (#51472)
> * Add scalar.conj() (#46596)
> * Add Tensor.copy_() for ComplexHalf tensors (#45339)

**Comment by iamlucaswolf** [comment]
That's great news. I hope to look into it over the weekend!

**Comment by josh146** [comment]
Can confirm that PyTorch 1.8 seems to support differentiating complex matmul! 🎉 

```python
        def circuit(x):
            I = torch.tensor([[1, 0], [0, 1]], dtype=torch.complex128)
            X = torch.tensor([[0, 1], [1, 0]], dtype=torch.complex128)
            Y = torch.tensor([[0j, -1j], [1j, 0j]], dtype=torch.complex128)
            Z = torch.tensor([[1, 0], [0, -1]], dtype=torch.complex128)
            init_state = torch.tensor([0.0, 1.0], dtype=torch.complex128)

            RY = torch.cos(x[0] / 2) * I + 1j * torch.sin(-x[0] / 2) * Y
            RX = torch.cos(x[1] / 2) * I + 1j * torch.sin(-x[1] / 2) * X

            state = RX @ RY @ init_state
            return torch.real(torch.conj(state) @ Z @ state)


        x = torch.tensor([1.0, 2.0], requires_grad=True)
        res = circuit(x)
        res.backward()
        g = x.grad
```

**Comment by antalszava** [comment]
With #1360 merged, closing this.

**Pull request closed without merging**